File: app/src/main/java/com/wishnewjam/dubstepfm/MainService.java
package com.wishnewjam.dubstepfm;

import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.wifi.WifiManager;
import android.os.Binder;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.PowerManager;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;
import android.view.KeyEvent;

import java.io.IOException;
import java.util.Iterator;
import java.util.Set;

public class MainService extends Service implements MediaPlayer.OnPreparedListener, MediaPlayer.OnCompletionListener, MediaPlayer.OnErrorListener {

    private static final int NOTIFICATION_ID = 43432;
    public static final long TIME_TO_RETRY = 3000;

    public static final String ACTION_PLAY = "com.example.action.PLAY";
    public static final String WAKE_LOCK = "mp_wakelock";
    public static final String SP_KEY_BITRATE = "link";

    public static final String PLAYING = "playing";
    public static final String WAITING = "waiting";
    public static final String ERROR = "error";

    MediaPlayer mMediaPlayer = null;
    private WifiManager.WifiLock wifiLock;
    private String currentLink;
    private Statuses currentStatus = Statuses.IDLE;
    private AudioManager audioManager;
    private static Handler retryHandler;
    private MediaButtonIntentReceiver mMediaButtonReceiver;

    @Override
    public void onCompletion(MediaPlayer mp) {
        if (retryHandler != null) {
            retryHandler.postDelayed(retryRunnable, TIME_TO_RETRY);
        }
    }

    @Override
    public boolean onError(MediaPlayer mp, int what, int extra) {
        sendMessage(ERROR);
        stopSelf();
        return false;
    }

    public enum Statuses {
        PLAYING, WAITING, IDLE
    }

    public MainService() {
    }


    public int onStartCommand(Intent intent, int flags, int startId) {
        if (retryHandler == null) {
            retryHandler = new Handler();
        }
        try {
            initMediaPlayer();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return START_STICKY;
    }


    private void initMediaPlayer() throws IOException {
        if (currentStatus == Statuses.WAITING) return;
        currentStatus = Statuses.WAITING;
        sendMessage(WAITING);
        audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        String link = PreferenceManager.getDefaultSharedPreferences(this)
                .getString(SP_KEY_BITRATE, Links.LINK_128);

        if (mMediaPlayer != null
                && mMediaPlayer.isPlaying()) {
            if (link.equals(currentLink)) {
                sendMessage(PLAYING);
                return;
            } else {
                mMediaPlayer.stop();
                mMediaPlayer.reset();
                mMediaPlayer = null;
            }
        }
        initHeadsetReceiver();

        currentLink = link;

        mMediaPlayer = new MediaPlayer();
        mMediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);

        mMediaPlayer.setDataSource(link);

        mMediaPlayer.setOnPreparedListener(this);
        mMediaPlayer.setOnCompletionListener(this);
        mMediaPlayer.setOnErrorListener(this);
        mMediaPlayer.setWakeMode(getApplicationContext(), PowerManager.PARTIAL_WAKE_LOCK);
        mMediaPlayer.prepareAsync();

    }

    private void initHeadsetReceiver() {
        if (mMediaButtonReceiver != null) {
            unregisterReceiver(mMediaButtonReceiver);
        }
        mMediaButtonReceiver = new MediaButtonIntentReceiver();
        IntentFilter mediaFilter = new IntentFilter(Intent.ACTION_MEDIA_BUTTON);
        mediaFilter.setPriority(IntentFilter.SYSTEM_HIGH_PRIORITY);
        registerReceiver(mMediaButtonReceiver, mediaFilter);

    }

    private final Runnable retryRunnable = new Runnable() {
        @Override
        public void run() {
            try {
                initMediaPlayer();
            } catch (IOException e) {
                if (retryHandler != null) {
                    retryHandler.postDelayed(retryRunnable, 10000);
                }
            }
        }
    };

    private void initWakeLock() {
        wifiLock = ((WifiManager) getSystemService(Context.WIFI_SERVICE))
                .createWifiLock(WifiManager.WIFI_MODE_FULL, WAKE_LOCK);
        wifiLock.acquire();
    }

    private void releaseWakeLock() {
        if (wifiLock != null) {
            wifiLock.release();
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return mBinder;
    }

    private final IBinder mBinder = new LocalBinder();

    public Statuses getStatus() {
        return currentStatus;
    }

    public class LocalBinder extends Binder {
        MainService getService() {
            return MainService.this;
        }
    }


    public void onPrepared(MediaPlayer player) {

        int result = audioManager.requestAudioFocus(afChangeListener,
                AudioManager.STREAM_MUSIC,
                AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK);

        if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            player.start();
            initNotification();
            initWakeLock();
            currentStatus = Statuses.PLAYING;
            sendMessage(PLAYING);
        }

    }

    private AudioManager.OnAudioFocusChangeListener afChangeListener
            = new AudioManager.OnAudioFocusChangeListener() {
        @Override
        public void onAudioFocusChange(int focusChange) {
            switch (focusChange) {
                case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                    if (mMediaPlayer != null) mMediaPlayer.pause();
                    break;
                case AudioManager.AUDIOFOCUS_GAIN:
                    if (mMediaPlayer != null) mMediaPlayer.start();
                    else {
                        try {
                            initMediaPlayer();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                    break;
                case AudioManager.AUDIOFOCUS_LOSS:
                    if (mMediaPlayer != null) {
                        if (mMediaPlayer.isPlaying()) {
                            mMediaPlayer.stop();
                        }
                        mMediaPlayer.reset();
                        mMediaPlayer.release();
                        mMediaPlayer = null;

                    } else {
                        stopSelf();
                    }
                    break;

            }


            if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT) {
                if (mMediaPlayer != null) mMediaPlayer.pause();
            } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
                try {
                    initMediaPlayer();
                } catch (IOException e) {
                    e.printStackTrace();
                    stopSelf();
                }
            } else if (focusChange == AudioManager.AUDIOFOCUS_LOSS) {
                stopSelf();
            }

        }
    };

    private void initNotification() {

        Intent resultIntent = new Intent(this, MainActivity.class);

        PendingIntent resultPendingIntent =
                PendingIntent.getActivity(
                        this,
                        0,
                        resultIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );

        NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(this)
                        .setSmallIcon(R.drawable.ic_notification)
                        .setContentTitle(getString(R.string.notification_title))
                        .setContentText(getString(R.string.notification_text))
                        .setContentIntent(resultPendingIntent);

        startForeground(NOTIFICATION_ID, mBuilder.build());
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mMediaButtonReceiver != null) {
            unregisterReceiver(mMediaButtonReceiver);
        }
        if (audioManager != null) {
            audioManager.abandonAudioFocus(afChangeListener);
        }
        releaseWakeLock();
        if (mMediaPlayer != null) {
            if (mMediaPlayer.isPlaying()) {
                mMediaPlayer.stop();
            }
            mMediaPlayer.reset();
            mMediaPlayer.release();
            mMediaPlayer = null;
        }
        if (retryHandler != null) {
            retryHandler.removeCallbacks(retryRunnable);
        }
    }

    private void sendMessage(String message) {
        Intent intent = new Intent(MainActivity.ACTION_RECEIVE_AUDIO_INFO);
        intent.putExtra(MainActivity.TAG_MESSAGE, message);
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
    }


    public class MediaButtonIntentReceiver extends BroadcastReceiver {
        private static final String LOG_TAG = "Media_receiver";

        @Override
        public void onReceive(Context context, Intent intent) {
            String intentAction = intent.getAction();
            if (!Intent.ACTION_MEDIA_BUTTON.equals(intentAction)) {
                return;
            }
            KeyEvent event = intent.getParcelableExtra(Intent.EXTRA_KEY_EVENT);
            if (event == null) {
                return;
            }
            int action = event.getAction();
            if (action == KeyEvent.ACTION_DOWN) {
                if (event.getKeyCode() == KeyEvent.KEYCODE_MEDIA_PAUSE) {
                    if (mMediaPlayer != null) {
                        if (mMediaPlayer.isPlaying()) {
                            mMediaPlayer.pause();
                        } else {
                            mMediaPlayer.prepareAsync();
                        }
                    }
                }

                if (event.getKeyCode() == KeyEvent.KEYCODE_MEDIA_PLAY) {
                    if (mMediaPlayer != null) {
                        mMediaPlayer.prepareAsync();
                    }
                }
            }
            abortBroadcast();

        }

    }

}

File: app/src/main/java/com/wishnewjam/dubstepfm/MainService.kt
package com.wishnewjam.dubstepfm

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Build
import android.os.Handler
import androidx.annotation.OptIn
import androidx.core.app.NotificationCompat
import androidx.media3.common.MediaMetadata
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.common.util.UnstableApi
import androidx.media3.session.MediaSession
import androidx.media3.session.MediaSessionService
import com.wishnewjam.dubstepfm.Tools.logDebug

@UnstableApi
class MainService : MediaSessionService() {

    companion object {
        private const val NOTIFICATION_ID = 43432
        const val SP_KEY_BITRATE = "link"
        const val MAX_ERROR_ATTEMPTS = 10

        const val NOTIFICATION_STATUS_PLAY = 1
        const val NOTIFICATION_STATUS_STOP = 0
        const val NOTIFICATION_STATUS_CONNECTING = 2
        const val NOTIFICATION_STATUS_LOADING = 3
        const val NOTIFICATION_STATUS_ERROR = 4
    }

    private lateinit var mediaPlayerInstance: MediaPlayerInstance
    private lateinit var mediaSession: MediaSession
    private val handler = Handler()

    private var notificationStatus = NOTIFICATION_STATUS_STOP
    private var errorPlayAttempts: Int = 0

    override fun onCreate() {
        super.onCreate()
        mediaPlayerInstance = (application as MyApplication).mediaPlayerInstance
        val player = mediaPlayerInstance.getPlayer()

        mediaSession = MediaSession.Builder(this, player)
            .setCallback(SessionCallback())
            .build()

        // Add player listener
        player.addListener(playerListener)
    }

    override fun onGetSession(controllerInfo: MediaSession.ControllerInfo): MediaSession? {
        return mediaSession
    }

    override fun onDestroy() {
        super.onDestroy()
        try {
            mediaSession.player.removeListener(playerListener)
            mediaSession.release()
        } catch (e: Exception) {
            logDebug { "Exception onDestroy: ${e.message}" }
        }
    }

    private val playerListener = object : Player.Listener {
        override fun onPlaybackStateChanged(playbackState: Int) {
            super.onPlaybackStateChanged(playbackState)
            logDebug { "PlayerListener: onPlaybackStateChanged = $playbackState" }
            updateNotificationBasedOnPlayerState()
        }

        override fun onMediaMetadataChanged(mediaMetadata: MediaMetadata) {
            super.onMediaMetadataChanged(mediaMetadata)
            logDebug { "PlayerListener: onMediaMetadataChanged = ${mediaMetadata.title}" }
            buildNotification(notificationStatus)
        }

        override fun onPlayerError(error: PlaybackException) {
            super.onPlayerError(error)
            logDebug { "PlayerListener: onPlayerError = $error" }
            handleError()
        }

        override fun onIsPlayingChanged(isPlaying: Boolean) {
            super.onIsPlayingChanged(isPlaying)
            logDebug { "PlayerListener: onIsPlayingChanged = $isPlaying" }
            if (isPlaying) {
                errorPlayAttempts = 0
            }
            updateNotificationBasedOnPlayerState()
        }
    }

    private fun updateNotificationBasedOnPlayerState() {
        val player = mediaSession.player
        val state = player.playbackState
        val isPlaying = player.isPlaying

        when {
            state == Player.STATE_BUFFERING -> {
                buildNotification(NOTIFICATION_STATUS_LOADING)
            }

            isPlaying -> {
                buildNotification(NOTIFICATION_STATUS_PLAY)
            }

            state == Player.STATE_IDLE || state == Player.STATE_ENDED -> {
                buildNotification(NOTIFICATION_STATUS_STOP)
            }

            else -> {
                buildNotification(NOTIFICATION_STATUS_STOP)
            }
        }
    }

    @OptIn(UnstableApi::class)
    private fun handleError() {
        if (++errorPlayAttempts < MAX_ERROR_ATTEMPTS) {
            buildNotification(NOTIFICATION_STATUS_CONNECTING)
            handler.postDelayed({ mediaPlayerInstance.callPlay() }, 10000)
        } else {
            buildNotification(NOTIFICATION_STATUS_ERROR)
        }
    }

    private fun buildNotification(keyCode: Int?) {
        val player = mediaSession.player
        val metadata = player.mediaMetadata
        val title = metadata.title ?: ""
        val subtitle = metadata.artist ?: ""
        notificationStatus = keyCode ?: NOTIFICATION_STATUS_STOP

        val activityIntent = Intent(applicationContext, MainActivity::class.java)
        val activityPendingIntent = PendingIntent.getActivity(
            applicationContext, 0, activityIntent,
            PendingIntent.FLAG_IMMUTABLE
        )

        val builder = NotificationCompat.Builder(applicationContext, "default")
            .setContentTitle(title)
            .setContentText(subtitle)
            .setContentIntent(activityPendingIntent)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setSmallIcon(R.drawable.ic_notification)
            .setColor(Color.BLACK)

        // If you have the androidx.media dependency, you can use MediaStyle:
        // implementation "androidx.media:media:1.x.x"
//        builder.setStyle(
//            androidx.media.app.NotificationCompat.MediaStyle()
//                .setMediaSession(mediaSession.sessionCompatToken)
//                .setShowActionsInCompactView(0)
//                .setShowCancelButton(true)
//                .setCancelButtonIntent(buildStopPendingIntent())
//        )

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP) {
            try {
                val bitmap = BitmapFactory.decodeResource(resources, R.drawable.logo_screen)
                builder.setLargeIcon(bitmap)
            } catch (e: Exception) {
                logDebug { "exception while decoding logo img: ${e.message}" }
            }
        }

        val notificationManager =
            getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        initChannels(notificationManager)

        val notification = when (notificationStatus) {
            NOTIFICATION_STATUS_PLAY -> {
                builder.addAction(
                    R.drawable.ic_stop,
                    getString(R.string.stop),
                    buildStopPendingIntent()
                )
                builder.build().also { startForeground(NOTIFICATION_ID, it) }
            }
            NOTIFICATION_STATUS_CONNECTING -> {
                builder.setContentTitle(getString(R.string.connecting))
                builder.setContentText("")
                builder.addAction(
                    R.drawable.ic_stop,
                    getString(R.string.stop),
                    buildStopPendingIntent()
                )
                builder.build().also { startForeground(NOTIFICATION_ID, it) }
            }

            NOTIFICATION_STATUS_STOP -> {
                builder.addAction(
                    R.drawable.ic_play,
                    getString(R.string.play),
                    buildPlayPendingIntent()
                )
                val notif = builder.build()
                notificationManager.notify(NOTIFICATION_ID, notif)
                stopForeground(false)
                notif
            }

            NOTIFICATION_STATUS_LOADING -> {
                builder.setContentTitle(getString(R.string.loading))
                builder.setContentText("")
                builder.addAction(
                    R.drawable.ic_stop,
                    getString(R.string.stop),
                    buildStopPendingIntent()
                )
                builder.build().also { startForeground(NOTIFICATION_ID, it) }
            }

            NOTIFICATION_STATUS_ERROR -> {
                builder.addAction(
                    R.drawable.ic_play,
                    getString(R.string.play),
                    buildPlayPendingIntent()
                )
                val notif = builder.build()
                notificationManager.notify(NOTIFICATION_ID, notif)
                stopForeground(false)
                notif
            }

            else -> {
                builder.addAction(
                    R.drawable.ic_play,
                    getString(R.string.play),
                    buildPlayPendingIntent()
                )
                val notif = builder.build()
                notificationManager.notify(NOTIFICATION_ID, notif)
                stopForeground(false)
                notif
            }
        }
    }

    private fun buildPlayPendingIntent(): PendingIntent {
        // We can define a custom action intent, or rely on the UI sending player commands.
        // For simplicity, let's just start MainActivity. The user can press play in the UI.
        // If you need a direct command, consider a SessionCommand and onCustomCommand().
        val playIntent = Intent(applicationContext, MainActivity::class.java)
        return PendingIntent.getActivity(
            applicationContext,
            0,
            playIntent,
            PendingIntent.FLAG_IMMUTABLE
        )
    }

    private fun buildStopPendingIntent(): PendingIntent {
        // Similarly, just start MainActivity or handle via custom command if needed.
        // Or handle via media button event if user stops from the notification.
        val stopIntent = Intent(applicationContext, MainService::class.java)
        stopIntent.action = "ACTION_STOP"
        return PendingIntent.getService(
            applicationContext,
            0,
            stopIntent,
            PendingIntent.FLAG_IMMUTABLE
        )
    }

    private fun initChannels(notificationManager: NotificationManager) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val channel =
            NotificationChannel("default", "DUBSTEP.FM", NotificationManager.IMPORTANCE_LOW)
        notificationManager.createNotificationChannel(channel)
    }

    @UnstableApi
    inner class SessionCallback : MediaSession.Callback {
        @OptIn(UnstableApi::class)
        override fun onConnect(
            session: MediaSession,
            controller: MediaSession.ControllerInfo
        ): MediaSession.ConnectionResult {
            // Accept all controllers.
            return MediaSession.ConnectionResult.AcceptedResultBuilder(session).build()
        }

        override fun onPostConnect(session: MediaSession, controller: MediaSession.ControllerInfo) {
            logDebug { "SessionCallback: onPostConnect from ${controller.packageName}" }
        }

        override fun onDisconnected(
            session: MediaSession,
            controller: MediaSession.ControllerInfo
        ) {
            logDebug { "SessionCallback: onDisconnected ${controller.packageName}" }
        }

        override fun onMediaButtonEvent(
            session: MediaSession,
            controller: MediaSession.ControllerInfo,
            intent: Intent
        ): Boolean {
            // Handle media button events like play/pause/stop from headset or Bluetooth device.
            val player = session.player
            val keyEvent = intent.getParcelableExtra<android.view.KeyEvent>(Intent.EXTRA_KEY_EVENT)
            if (keyEvent != null && keyEvent.action == android.view.KeyEvent.ACTION_DOWN) {
                when (keyEvent.keyCode) {
                    android.view.KeyEvent.KEYCODE_MEDIA_STOP -> {
                        player.stop()
                        buildNotification(NOTIFICATION_STATUS_STOP)
                        return true
                    }

                    android.view.KeyEvent.KEYCODE_MEDIA_PLAY -> {
                        player.play()
                        buildNotification(NOTIFICATION_STATUS_PLAY)
                        return true
                    }

                    android.view.KeyEvent.KEYCODE_HEADSETHOOK -> {
                        if (player.isPlaying) {
                            player.stop()
                            buildNotification(NOTIFICATION_STATUS_STOP)
                        } else {
                            player.play()
                            buildNotification(NOTIFICATION_STATUS_PLAY)
                        }
                        return true
                    }
                }
            }
            return false
        }
    }
}
